### PR TITLE
PPTP-1880: Deregistration

### PR DESCRIPTION
**Pull request by RyanKellyHMRC**
New endpoint to support deregistration.

#### Check list 
 - [x] `./precheck` was executed (Integration/Component/Unit tests)
 - [x] `sbt scalafmt test:scalafmt` was executed
 - [ ] Required Environment Config has been amended/added - N/A
 - [ ] Links to dependencies have been included (BE/FE work) - N/A
 - [ ] User Acceptance Tests (UAT) were run locally and have passed - N/A

### CODE REVIEW

File: conf/app.routes [hunk 1]
```diff
@@ -13,6 +13,8 @@ POST        /subscriptions/:safeNumber           uk.gov.hmrc.plasticpackagingtax
 GET         /subscriptions/:pptReference   uk.gov.hmrc.plasticpackagingtaxregistration.controllers.SubscriptionController.get(pptReference: String)
 PUT         /subscriptions/:pptReference   uk.gov.hmrc.plasticpackagingtaxregistration.controllers.SubscriptionController.update(pptReference: String)
 
+DELETE      /subscriptions/:pptReference   uk.gov.hmrc.plasticpackagingtaxregistration.controllers.DeregistrationController.deregister(pptReference: String)
```
Comment (paulblakehmrc) [inline conf/app.routes:16]: arguably it's a PUT as the record still exists in the backend

File: app/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/DeregistrationController.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    subscription: Subscription,
+    deregistrationReason: DeregistrationReason
+  ) =
+    subscription.copy(changeOfCircumstanceDetails =
+      Some(
+        ChangeOfCircumstanceDetails(changeOfCircumstance = "Deregistration",
+                                    deregistrationDetails = Some(
+                                      DeregistrationDetails(
+                                        deregistrationReason = deregistrationReason.toString,
+                                        deregistrationDate = LocalDate.now().toString,
```
Comment (paulblakehmrc) [inline app/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/DeregistrationController.scala:75]: do we want to start using `LocalDateTime.now(ZoneOffset.UTC)`?

**Review by paulblakehmrc — APPROVED** [review]

File: app/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/DeregistrationController.scala (hunk 2 above)
Comment (RyanKellyHMRC) [reply app/uk/gov/hmrc/plasticpackagingtaxregistration/controllers/DeregistrationController.scala:75]: Yeah, that's probably safer. I'll amend there.

File: conf/app.routes (hunk 1 above)
Comment (RyanKellyHMRC) [reply conf/app.routes:16]: That's implementation detail. The intent is to delete the subscription.

Comment (paulblakehmrc) [reply conf/app.routes:16]: It's more a request to delete. The outcome might be a delete or not.

Comment (ajchilts) [reply conf/app.routes:16]: If the sole responsibility is to delete a record then I would have used DELETE. If however there are a number of things going on behind the endpoint and delete is a side effect that may or may not happen I would have used a PUT.

**Review by ajchilts — APPROVED** [review]

File: conf/app.routes (hunk 1 above)
Comment (RyanKellyHMRC) [reply conf/app.routes:16]: Always controversial. We'll go with DELETE for now.

**Pull request merged**